### ISSUE TITLE
CPP Operator pointers

### ISSUE THREAD

**Issue by jlp765** [nim-lang/c2nim#60]
The following fails with
`Error: identifier expected, but found '*'`

for `c2nim --cpp` conversion of

```
class failClass1 {
public:
  int tmpI;
  operator MyOp1() { return tmpI; }
  operator MyOp2*() { return *tmpI; }   // <- this fails, for MyOp2*()
};
```

**Comment by Araq** [comment]
What does that even mean?  Implicit conversion to 'ptr to myOp2'?

**Comment by jlp765** [comment]
I'm no c++ guru, so I may not have described it correctly, but c2nim didn't like
``operator xxxxx*()``
and it is the ``*`` it doesn't like.  

I came across this in the FLTK library

```
class Fl_XFont_On_Demand
{
public:
  Fl_XFont_On_Demand(XFontStruct* p = NULL) : ptr(p) { }
  Fl_XFont_On_Demand& operator=(const Fl_XFont_On_Demand& x)
  { ptr = x.ptr;  return *this; }
  Fl_XFont_On_Demand& operator=(XFontStruct* p)
  { ptr = p;  return *this; }
  XFontStruct* value();
  operator XFontStruct*() { return value(); }
  XFontStruct& operator*() { return *value(); }
  XFontStruct* operator->() { return value(); }
  bool operator==(const Fl_XFont_On_Demand& x) { return ptr == x.ptr; }
  bool operator!=(const Fl_XFont_On_Demand& x) { return ptr != x.ptr; }
private:
  XFontStruct *ptr;
};
```

**Comment by jjzazuet** [comment]
HHHmmm... I don't think `c2nim` would actually be expected to understand that.

@Araq correct me if I'm wring but, `c2nim` only supports translation of a subset of ANSI C, right?

Thanks!

**Comment by Araq** [comment]
> c2nim only supports translation of a subset of ANSI C, right?

A subset of C++, that is. And if people can tell me what ``operator foo*`` does in C++, I can think about supporting it.

**Comment by cjhanks** [comment]
Typically `operator*` is paired with `operator->` when a class is responsible for managing internal data; `unique_ptr`, `shared_ptr`.

It typically means dereference the internal point and pass me the reference.

Example:
``` cpp

class Example {
public:
  struct Element {
    int a;
    int b;
  };

  Element*
  operator->() {
    return &e;
  }

  Element&
  operator*() {
    return e;
  }

private:
      Element e;
};


void
Function(Example::Element& element) {
  element.a = 4;
}

int
main()
{
  Example example;
  example->a = 3;

  Function(*example);
}
```

**Comment by Araq** [comment]
That's ``operator*``. I asked about ``operator namehere*``.

**Comment by cjhanks** [comment]
Ah, my misread.

`operator AnyTypeHere` is a cast operator.  It defines an implicit (or in C++11 explicit) type conversion from `this` to another type.

``` cpp
class TypeA {
public:
  int value;
};

class TypeB {
public:
  operator TypeA&() {
    return a;
  }

  operator TypeA*() {
    return &a;
  }

  // C++11 or later
  explicit operator int&() {
    return a.value;
  }

private:
  TypeA a;
};


void Function1(TypeA v) { }
void Function2(TypeA* v) { }
void Function3(TypeA v) { }
void Function4(int& v) { }

int
main()
{
  TypeB b;

  // implicit casts are 'okay'
  Function1(b); // calls operator::TypeA()
  Function2(b); // calls operator::TypeA*()
  Function3(b); // calls operator::TypeA()

  // explicit cast is required here
  Function4(static_cast<int&>(b)); // calls operator::int&()
}
```

**Comment by BigEpsilon** [comment]
This can be closed as it been fixed by #102
